dipy/reconst/qtdmri.py

```diff
@@ -0,0 +1,1952 @@
+        for j in range(1, 2 + n / 2):
+            l = n + 2 - 2 * j
+            for m in range(-l, l + 1):
+                for o in range(0, time_order+1):
+                    index_matrix.append([j, l, m, o])
+
+    return np.array(index_matrix)
+
+
+def qtdmri_laplacian_reg_matrix(ind_mat, us, ut, S_mat, T_mat, U_mat,
+                                part1_ut_precomp=None,
+                                part23_ut_precomp=None,
+                                part4_ut_precomp=None):
+    """Computes the cartesian qt-dMRI Laplacian regularization matrix. If
+    given, uses precomputed matrices for temporal and spatial regularization
+    matrices to speed up computation. Follows the the formulation of Appendix B
+    in [1].
+    
+    References
+    ----------
+    .. [1] Fick, Rutger HJ, et al. "Non-Parametric GraphNet-Regularized
+        Representation of dMRI in Space and Time", Medical Image Analysis,
+        2017.
+    """
+    part1_us = mapmri.mapmri_laplacian_reg_matrix(ind_mat[:, :3], us,
+                                                  S_mat, T_mat, U_mat)
+    part23_us = part23_reg_matrix_q(ind_mat, U_mat, T_mat, us)
+    part4_us = part4_reg_matrix_q(ind_mat, U_mat, us)
+
+    if part1_ut_precomp is None:
+        part1_ut = part1_reg_matrix_tau(ind_mat, ut)
+    else:
+        part1_ut = part1_ut_precomp / ut
+    if part23_ut_precomp is None:
+        part23_ut = part23_reg_matrix_tau(ind_mat, ut)
+    else:
+        part23_ut = part23_ut_precomp * ut
+    if part4_ut_precomp is None:
+        part4_ut = part4_reg_matrix_tau(ind_mat, ut)
+    else:
+        part4_ut = part4_ut_precomp * ut ** 3
+
+    regularization_matrix = (
+        part1_us * part1_ut + part23_us * part23_ut + part4_us * part4_ut
+    )
+    return regularization_matrix
+
+
+def qtdmri_isotropic_laplacian_reg_matrix(ind_mat, us, ut,
+                                          part1_uq_iso_precomp=None,
+                                          part1_ut_precomp=None,
+                                          part23_ut_precomp=None,
+                                          part4_ut_precomp=None):
+    """Computes the spherical qt-dMRI Laplacian regularization matrix. If
+    given, uses precomputed matrices for temporal and spatial regularization
+    matrices to speed up computation. Follows the the formulation of Appendix C
+    in [1].
+    
+    References
+    ----------
+    .. [1] Fick, Rutger HJ, et al. "Non-Parametric GraphNet-Regularized
+        Representation of dMRI in Space and Time", Medical Image Analysis,
+        2017.
+    """
+    if part1_uq_iso_precomp is None:
+        part1_us = (
+            mapmri.mapmri_isotropic_laplacian_reg_matrix_from_index_matrix(
+                ind_mat[:, :3], us[0]
+            )
+        )
+    else:
+        part1_us = part1_uq_iso_precomp * us[0]
+
+    if part1_ut_precomp is None:
+        part1_ut = part1_reg_matrix_tau(ind_mat, ut)
+    else:
+        part1_ut = part1_ut_precomp / ut
```

Comment:
is this block really necessary ? Indeed, the variable "part1_ut" is not used. This variable is override L1471